src/lib/storage/dictionary_segment.hpp

```diff
@@ -0,0 +1,159 @@
+    return _dictionary;
+  }
+
+  // returns an underlying data structure
+  std::shared_ptr<BaseAttributeVector> attribute_vector() const {
+    return _attribute_vector;
+  }
+
+  // return the value represented by a given ValueID
+  const T& value_by_value_id(ValueID value_id) const {
+    return _dictionary->at(value_id);
+  }
+
+  // returns the first value ID that refers to a value >= the search value
+  // returns INVALID_VALUE_ID if all values are smaller than the search value
+  ValueID lower_bound(T value) const {
+    auto result_memory_address = std::lower_bound(_dictionary->begin(), _dictionary->end(), value);
+    auto result = static_cast<ValueID>(std::distance(_dictionary->begin(), result_memory_address));
+    if (result == _dictionary->size()) {
+      return INVALID_VALUE_ID;
+    }
+    return result;
+  }
+
+  // same as lower_bound(T), but accepts an AllTypeVariant
+  ValueID lower_bound(const AllTypeVariant& value) const {
+    return lower_bound(static_cast<T>(value));
+  }
+
+  // returns the first value ID that refers to a value > the search value
+  // returns INVALID_VALUE_ID if all values are smaller than or equal to the search value
+  ValueID upper_bound(T value) const {
+    auto result_memory_address = std::upper_bound(_dictionary->begin(), _dictionary->end(), value);
+    auto result = static_cast<ValueID>(std::distance(_dictionary->begin(), result_memory_address));
+    if (result == _dictionary->size()) {
+      return INVALID_VALUE_ID;
+    }
+    return result;
+  }
+
+  // same as upper_bound(T), but accepts an AllTypeVariant
+  ValueID upper_bound(const AllTypeVariant& value) const {
+    return upper_bound(static_cast<T>(value));
+  }
+
+  // return the number of unique_values (dictionary entries)
+  size_t unique_values_count() const {
+    return _dictionary->size();
+  }
+
+  // return the number of entries
+  size_t size() const override {
+    return _attribute_vector->size();
+  }
+
+  // returns the calculated memory usage of the segment
+  size_t estimate_memory_usage() const {
+    auto dic_size = _dictionary->size() * sizeof(T);
+    auto att_size = _attribute_vector->size() * _attribute_vector->width();
+
+    return dic_size + att_size;
+  }
+
+ protected:
+  std::shared_ptr<std::vector<T>> _dictionary;
+  std::shared_ptr<BaseAttributeVector> _attribute_vector;
+
+  // set the attribute vector with the current width and length depending on the incoming data
+  void _set_attribute_vector(const size_t dictionary_size, const size_t attribute_vector_size) {
+    if (dictionary_size <= std::numeric_limits<uint8_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint8_t>>(attribute_vector_size);
+    } else if (dictionary_size <= std::numeric_limits<uint16_t>::max()) {
+      _attribute_vector = std::make_shared<FixedSizeAttributeVector<uint16_t>>(attribute_vector_size);
+    } else {
+      DebugAssert(dictionary_size <= std::numeric_limits<uint32_t>::max(),
```

Comment:
Since initialization of the attribute vector isn't performance-critical, it might make sense to use a standard assert here.

Reply:
Could make sense, I'd have to think about this and I'm not willing to.

Won't Fix